pkg/settlement/swap/transaction/monitor.go

```diff
@@ -0,0 +1,251 @@
+// Copyright 2021 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package transaction
+
+import (
+	"context"
+	"errors"
+	"io"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethersphere/bee/pkg/logging"
+)
+
+var ErrTransactionCancelled = errors.New("transaction cancelled")
+var ErrMonitorClosed = errors.New("monitor closed")
+
+// Monitor is a nonce-based watcher for transaction confirmations.
+// Instead of watching transactions individually, the senders nonce is monitored and transactions are checked based on this.
+// The idea is that if the nonce is still lower than that of a pending transaction, there is no point in actually checking the transaction for a receipt.
+// At the same time if the nonce was already used and this was a few blocks ago we can reasonably assume that it will never confirm.
+
+type Monitor interface {
+	io.Closer
+	// WatchTransaction watches the transaction until either there is 1 confirmation or a competing transaction with cancellationDepth confirmations.
+	WatchTransaction(txHash common.Hash, nonce uint64) (chan types.Receipt, chan error, error)
+}
+type transactionMonitor struct {
+	lock       sync.Mutex
+	ctx        context.Context    // context which is used for all backend calls
+	cancelFunc context.CancelFunc // function to cancel the above context
+	wg         sync.WaitGroup
+
+	logger  logging.Logger
+	backend Backend
+	sender  common.Address // sender of transactions which this instance can monitor
+
+	pollingInterval   time.Duration // time between checking for new blocks
+	cancellationDepth uint64        // number of blocks until considering a tx cancellation final
+
+	watches    map[*transactionWatch]struct{} // active watches
+	watchAdded chan struct{}                  // channel to trigger instant pending check
+}
+
+type transactionWatch struct {
+	receiptC chan types.Receipt // channel to which the receipt will be written once available
+	errC     chan error         // error channel (primarily for cancelled transactions)
+
+	txHash common.Hash // hash of the transaction to watch
+	nonce  uint64      // nonce of the transaction to watch
+}
+
+func NewMonitor(logger logging.Logger, backend Backend, sender common.Address, pollingInterval time.Duration, cancellationDepth uint64) Monitor {
+	ctx, cancelFunc := context.WithCancel(context.Background())
```

Comment:
there is one shared loop for all watched transactions, so which context should it use for the `NonceAt` call which is not even related to any specific transaction? If it uses a context created inside `watchPending` and cancelled in its defer than shutdown might not work at all as it might get stuck waiting on `NonceAt` or `BlockNumber` calls.

Where do you foresee a deadlock happening in the current code?